resilience4j-spring/src/main/java/io/github/resilience4j/circuitbreaker/configure/CircuitBreakerAnnotationConfigScanner.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2019 Olov Andersson
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.github.resilience4j.circuitbreaker.configure;
+
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.config.BeanDefinition;
+import org.springframework.beans.factory.config.BeanFactoryPostProcessor;
+import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
+import org.springframework.core.Ordered;
+import org.springframework.core.annotation.AnnotationUtils;
+import org.springframework.core.env.ConfigurableEnvironment;
+import org.springframework.core.env.MapPropertySource;
+
+import io.github.resilience4j.circuitbreaker.annotation.CircuitBreaker;
+import io.github.resilience4j.common.circuitbreaker.configuration.CircuitBreakerConfigurationProperties.InstanceProperties;
+
+/**
+ * Scans all bean definitions for CircuitBreaker annotations and collects annotated configuration so
+ * it can be added as a PropertySource to the environment. This is done as a
+ * BeanFactoryPostProcessor so we can ensure that we have found these before actual bean
+ * instantiation begins.
+ * <p>
+ * Unless resilience4j-spring-boot/resilience4j-spring-boot2/
+ * resilience4j-spring-cloud/resilience4j-spring-cloud2 is used this class must be explicitly added
+ * to the spring context as a bean for CircuitBreaker annotation exception config to be considered.
+ */
+public class CircuitBreakerAnnotationConfigScanner implements BeanFactoryPostProcessor, Ordered {
+	private static final Logger LOG = LoggerFactory.getLogger(CircuitBreakerAnnotationConfigScanner.class);
+
+    private ConfigurableEnvironment configurableEnvironment;
+
+    private int order = LOWEST_PRECEDENCE;
+    
+	public CircuitBreakerAnnotationConfigScanner(ConfigurableEnvironment configurableEnvironment) {
+        this.configurableEnvironment = configurableEnvironment;
+	}
+	
+	public void setOrder(int order) {
```

Comment:
I think it is not needed. If we want to change order, it would be better to get from `CircuitBreakerProperties`

Reply:
I think the Ordered  inteface is needed so that this BeanFactoryPostProcessor is processed before others, or?

Reply:
Yes it is needed. But I mean `setOrder()` is not needed. Like `CircuitBreakerAspect`, I think we can set order with properties. 

https://github.com/resilience4j/resilience4j/blob/c9a6fa2d036b41f17e300aa5d43885ea7e3fc9ab/resilience4j-spring/src/main/java/io/github/resilience4j/circuitbreaker/configure/CircuitBreakerConfigurationProperties.java#L23

Reply:
A user should not be able to reconfigure it. Otherwise it doesn't work anymore.

Reply:
yeah right. Then, we can change like below and remove `setOrder()` method.

```java
@Override
public int getOrder() {
    return LOWEST_PRECEDENCE;
}
```

Reply:
I added the setOrder to add flexibility to the library. It might not be necessary, but BeanFactoryPostProcessors can do all sorts of stuff to your application context , and we cannot anticipate how every single use-case might affect one another. Many exising BeanFactoryPostProcessors have this ability. This way you give library users a bit of flexibility for instance in case there is some other BeanFactoryPostProcessor that they would want to affect this one. There are several built-in factories in Spring that has this flexibility. I would guess that 99 out of 100 users won't bother with this. Note that the default order is lowest, so it will get processed last.

We can't set the order with CircuitBreakerConfigurationProperties because this factory is in fact used to help populate those properties.

We could skip the order though, if I remember correctly, it works correctly without it. People who want to customize could then choose to subclass instead.

Reply:
Removed the Ordered interface as it wasn't really needed.